### DISCUSSION TITLE
Entity ruler, matching priority and lemmatisation issues

### DISCUSSION THREAD

**Discussion by lfoppiano** [explosion/spaCy#6309] [Help: Coding & Implementations] [answered]
I'm implementing a EntityRuler with tons of patterns and I'm wondering if there is a way to configure the ruler to match first the largest entities. 

For example if I have the following patterns: 

```json
{"label":"L","pattern":[{"LOWER":"Fabry"},{"LOWER":"disease"}],"id":"1"}
{"label":"L","pattern":[{"LEMMA":"Fabry"},{"LEMMA":"disease"}],"id":"1"}
{"label":"L","pattern":[{"LOWER":"disease"}],"id":"2"}
{"label":"L","pattern":[{"LEMMA":"disease"}],"id":"2"}
```

when I process this sentence: 
```
Fabry disease is important. fabry disease is not the same.
```

I get 

```json
[{"text": "Fabry disease", "type": "L", "offsetStart": 0, "offsetEnd": 13, "id":
"1"}, {"text": "disease", "type": "L", "offsetStart": 34, "offsetEnd":
41, "id": "2"}]
```

I was expecting that first to match the largest patterns and then the smaller. 
Another question, how to keep also overlapping matches?  

## Info about spaCy

* **spaCy version:** 2.3.2
* **Platform:** macOS-10.15.7-x86_64-i386-64bit
* **Python version:** 3.8.5

**Comment by adrianeboyd** [comment, marked as answer]
The `EntityRuler` is a wrapper around the `Matcher` and `PhraseMatcher`, so if you need more control of how overlapping matches are managed, you may want to use the `Matcher` directly instead of using the `EntityRuler`. As a starting point, you could have a look at `EntityRuler.__call__` to see how entities are matched and filtered.

The current filtering for `EntityRuler` should choose the earliest span for overlapping matches, and if multiple spans have the same start token, then the longest one is preferred. Wait, I remember that there was a bug here, so with the current version of the library (v2.3.2) you won't see this exact behavior. (We need to work on releasing v2.3.3 with this bug fix. I've been planning to wait until we can add support for python 3.9, but we may not want to wait too much longer.) Here's the current version with the bug fix:

https://github.com/explosion/spaCy/blob/260c29794a1caa70f8b0702c31fcfecad6bfdadc/spacy/pipeline/entityruler.py#L86-L122

A `Doc` does not allow you to store overlapping entities because the entities are stored internally as IOB tags and you can only store one tag per token. If you need to store overlapping spans, you can use custom `Doc` or `Token` extensions, see: https://spacy.io/usage/processing-pipelines#custom-components-attributes .

**Comment by lfoppiano** [comment]
Thanks @adrianeboyd for the detailed clarification! 

Indeed, the `__call__` leverages both Matcher and PhraseMatcher. I noticed a strange behaviour, for example, while debugging my example, and checking the results of 
```
 matches = list(self.matcher(doc)) + list(self.phrase_matcher(doc)) 
```

I get three entities from the Matcher: 

```
0 = {tuple: 3} (8610959670337998882, 0, 2)
1 = {tuple: 3} (16063251451996509537, 1, 2)
2 = {tuple: 3} (16063251451996509537, 6, 7)
``` 
and by looking at  the content : 
```
`[doc[x[1]:x[2]] for x in self.matcher(doc)]`
0 = {Span: 2} Fabry disease
1 = {Span: 1} disease
2 = {Span: 1} disease
```

so in the first sentence, there are two matches: "Fabry disease" and "disease", while the sentence only "disease" is matching. 
I wonder, is it normal?

**Comment by adrianeboyd** [comment]
You probably want to use `LOWER` instead of `LEMMA` for terms like `Fabry` for this pattern. `LEMMA` can be a bit unpredictable because it depends on whether the tagger thinks it's a proper noun or not. If it thinks it's a proper noun, the lemma is the same as the word, but if it thinks it's a common noun, the lemma is lowercased. You can also use something like `{"LOWER": {"IN": ["disease", "diseases"]}}` to handle plural forms if the tagger is too inconsistent for your data.

**Comment by lfoppiano** [comment]
Understood. Thanks. Indeed it makes sense.

**Comment by lfoppiano** [comment]
> You probably want to use `LOWER` instead of `LEMMA` for terms like `Fabry` for this pattern. `LEMMA` can be a bit unpredictable because it depends on whether the tagger thinks it's a proper noun or not. If it thinks it's a proper noun, the lemma is the same as the word, but if it thinks it's a common noun, the lemma is lowercased. You can also use something like `{"LOWER": {"IN": ["disease", "diseases"]}}` to handle plural forms if the tagger is too inconsistent for your data.

I'm coming back on this, as the "LEMMA" confusion is haunting me down 😁 


I have some cases that I still don't understand...  
Why `tumours` as `NOUN` has the same lemma `tumours` but this does not happen for when the POS think is a `VERB`?


Case 1: 
```
self.nlp("the brain tumours")[2].lemma_
tumour

self.nlp("the brain tumours")[2].pos_
VERB
```

Case 2: 
```
self.nlp("the brain tumours is")[2].lemma_
tumours

self.nlp("the brain tumours is")[2].pos_
'NOUN'
```

Beside that, I'm now trying to find a solution to get rid of all the LEMMAs in the patterns, but it is not so easy: 
 - assuming the string I want to add as patterns, are mixed with singular, plural and various inflections, I should find the stem/root and re-constrcut all the inflections which can be added as you mentioned before with the `LOWER-IN []` pattern. 
 - inflections seems to be related to the type POS tag of the words, but let's say we generate them for all - is there a way in spaCy? 

If you have some hint or suggestion that I could look up to solve this problem, that would be great. 

Thank you in advance 
Luca

**Comment by adrianeboyd** [comment]
The rule-based lemmatizer just probably isn't a good match for this task because of the orthographic variants and low tagging accuracy for tokens like `BRAIN`.

If you want lemmas in some form, you could switch to the lookup lemmas, which will just use the token text and not any other information.

Ideally you'd just remove the `tagger` from the pipeline to get the lookup lemmas, but this is also not as straightforward as it should be with the `Matcher` because the lookup lemmas aren't stored on the token data internally by default, just provided as a backup through python API, so you have to add them manually so they're availabe in the token data (you can do this with a custom pipeline component):

```python
for token in doc:
    token.lemma_ = nlp.vocab.morphology.lemmatizer.lookup(token.text)
```

You could also convert your tokens to the lowercase before lemmatizing to normalize a bit further (since things like `BRAIN` aren't going to be in the lookup table):

```
for token in doc:
    token.lemma_ = nlp.vocab.morphology.lemmatizer.lookup(token.lower_)
```

Then you will have more consistent lemmatized forms that are easier to match.

**Comment by lfoppiano** [comment]
Thanks! I will dig in this direction then ;-)
